### PR TITLE
Introduces Address type to be used in secondary IPv4 and IPv6 inspect data structure

### LINKED ISSUE #12165: docker api: NetworkSettings.SecondaryIPv6Addresses type mismatch
**Is this a BUG REPORT or FEATURE REQUEST? **

/kind bug

**Description**

When inspecting a container with non-empty `NetworkSettings.SecondaryIPv6Addresses` I get the following error: `Error response from daemon: json: cannot unmarshal string into Go struct field NetworkSettings.SecondaryIPv6Addresses of type network.Address` 

**Steps to reproduce the issue:**

Having created a `container` with non-empty `NetworkSettings.SecondaryIPv6Addresses` call podman with the docker SDK like this:
```golang
inspected, err := cli.ContainerInspect(context.Background(), container.ID)                                                                                                                  
if err != nil {                                                                                                                                                                             
  log.Panicf("could not inspect container %v\n", err)                                                                                                                                       
} 
```
**Additional information you deem important (e.g. issue happens only occasionally):**
The `[]string` type defined here https://github.com/containers/podman/blob/056f492f59c333d521ebbbe186abde0278e815db/libpod/define/container_inspect.go#L561 is not compatible with `[]networktypes.Address` defined here https://github.com/moby/moby/blob/7b9275c0da707b030e62c96b679a976f31f929d3/daemon/network/settings.go#L25.  (networktypes.Address from https://github.com/moby/moby/blob/7b9275c0da707b030e62c96b679a976f31f929d3/api/types/network/network.go#L7-L10)

**Output of `podman version`:**

```
Version:      3.4.1
API Version:  3.4.1
Go Version:   go1.16.8
Built:        Wed Oct 20 14:35:28 2021
OS/Arch:      linux/amd64
```

**Output of `podman info --debug`:**

```
host:
  arch: amd64
  buildahVersion: 1.23.1
  cgroupControllers:
  - cpuset
  - cpu
  - cpuacct
  - blkio
  - memory
  - devices
  - freezer
  - net_cls
  - perf_event
  - net_prio
  - hugetlb
  - pids
  - misc
  cgroupManager: cgroupfs
  cgroupVersion: v1
  conmon:
    package: conmon-2.0.30-2.fc34.x86_64
    path: /usr/bin/conmon
    version: 'conmon version 2.0.30, commit: '
  cpus: 8
  distribution:
    distribution: fedora
    variant: container
    version: "34"
  eventLogger: file
  hostname: localhost.localdomain
  idMappings:
    gidmap: null
    uidmap: null
  kernel: 5.14.14-200.fc34.x86_64
  linkmode: dynamic
  logDriver: k8s-file
  memFree: 19763777536
  memTotal: 33391837184
  ociRuntime:
    name: crun
    package: crun-1.2-1.fc34.x86_64
    path: /usr/bin/crun
    version: |-
      crun version 1.2
      commit: 4f6c8e0583c679bfee6a899c05ac6b916022561b
      spec: 1.0.0
      +SYSTEMD +SELINUX +APPARMOR +CAP +SECCOMP +EBPF +CRIU +YAJL
  os: linux
  remoteSocket:
    path: /run/podman/podman.sock
  security:
    apparmorEnabled: false
    capabilities: CAP_CHOWN,CAP_DAC_OVERRIDE,CAP_FOWNER,CAP_FSETID,CAP_KILL,CAP_NET_BIND_SERVICE,CAP_SETFCAP,CAP_SETGID,CAP_SETPCAP,CAP_SETUID,CAP_SYS_CHROOT
    rootless: false
    seccompEnabled: true
    seccompProfilePath: /usr/share/containers/seccomp.json
    selinuxEnabled: false
  serviceIsRemote: false
  slirp4netns:
    executable: /usr/bin/slirp4netns
    package: slirp4netns-1.1.12-2.fc34.x86_64
    version: |-
      slirp4netns version 1.1.12
      commit: 7a104a101aa3278a2152351a082a6df71f57c9a3
      libslirp: 4.4.0
      SLIRP_CONFIG_VERSION_MAX: 3
      libseccomp: 2.5.0
  swapFree: 8589930496
  swapTotal: 8589930496
  uptime: 25m 10.97s
plugins:
  log:
  - k8s-file
  - none
  - journald
  network:
  - bridge
  - macvlan
  volume:
  - local
registries:
  search:
  - registry.fedoraproject.org
  - registry.access.redhat.com
  - docker.io
  - quay.io
store:
  configFile: /etc/containers/storage.conf
  containerStore:
    number: 0
    paused: 0
    running: 0
    stopped: 0
  graphDriverName: overlay
  graphOptions:
    overlay.mountopt: nodev,metacopy=on
  graphRoot: /var/lib/containers/storage
  graphStatus:
    Backing Filesystem: btrfs
    Native Overlay Diff: "false"
    Supports d_type: "true"
    Using metacopy: "true"
  imageStore:
    number: 0
  runRoot: /run/containers/storage
  volumePath: /var/lib/containers/storage/volumes
version:
  APIVersion: 3.4.1
  Built: 1634740528
  BuiltTime: Wed Oct 20 14:35:28 2021
  GitCommit: ""
  GoVersion: go1.16.8
  OsArch: linux/amd64
  Version: 3.4.1
```

**Package info :**
```
podman-3.4.1-1.fc34.x86_64
```

**Have you tested with the latest version of Podman and have you checked the Podman Troubleshooting Guide? (https://github.com/containers/podman/blob/master/troubleshooting.md)**

Yes

### PR DESCRIPTION
**Pull request by fgimenez**
#### What this PR does / why we need it:

Fixes discrepancy between the types used in inspect for docker and podman. This causes a panic when using the docker client against podman when the secondary IP fields in the `NetworkSettings` inspect field are populated.

#### How to verify it

* podman system service --log-level debug -t 0 unix:///${HOME}/podman.sock > /var/log/podman.log 2>&1 &
* Create a container with `SecondaryIPV6Address` populated in inspect's `NetworkSettings`
* ln -s /root/podman.sock /var/run/docker.sock
* Use docker cli to inspect the container:
```golang
inspected, err := cli.ContainerInspect(context.Background(), container.ID)
if err != nil {
  log.Panicf("could not inspect container %v\n", err)
} 
```
#### Which issue(s) this PR fixes:

Fixes #12165

Signed-off-by: Federico Gimenez <fgimenez@redhat.com>

### CODE REVIEW

**Comment by rhatdan** [comment]
@Luap99 @mheon @baude PTAL

**Comment by Luap99** [comment]
@fgimenez How do you create such a container?

**Comment by Luap99** [comment]
I think we might want to fix this only for the compat API and not regular inspect, @mheon WDYT?

**Comment by fgimenez** [comment]
> @fgimenez How do you create such a container?

I'm finding this issue on CI doing something similar to:
```
$ sudo podman run --privileged -ti --entrypoint /bin/bash quay.io/kubevirtci/bootstrap:podman-test16
# inside the container
$ podman system service --log-level debug -t 0 unix:///${HOME}/podman.sock > /var/log/podman.log 2>&1 &
$ git clone https://github.com/kubevirt/kubevirtci && cd kubevirtci
$ make cluster-up # fails with default podman 3.4.1, succeeds with the changes in this PR
$ podman inspect k8s-1.20-node01
[
    {
................
        "NetworkSettings": {
            "EndpointID": "",
            "Gateway": "10.88.0.1",
            "IPAddress": "10.88.0.2",
            "IPPrefixLen": 16,
            "IPv6Gateway": "",
            "GlobalIPv6Address": "",
            "GlobalIPv6PrefixLen": 0,
            "SecondaryIPv6Addresses": [
                "2001:db8:1::2/64"
            ],
..............
```
Will try to find a simpler example

**Comment by mheon** [comment]
Tend to agree with @Luap99 that this should be compat API only - this looks like a breaking change to me, and while we are in a breaking-change mode for 4.0, I don't see a good reason we should change inspect now, instead of just fixing the Docker-compat endpoint.

**Comment by fgimenez** [comment]
@Luap99 @mheon sgtm, where should I add the changes to make them compat API only?

**Comment by Luap99** [comment]
https://github.com/containers/podman/blob/c0351a75ae1e4a74a633ac95f1d2ca9c7d96f044/pkg/api/handlers/compat/containers.go#L350-L357
This is where the error happens. We cannot use json.marshal/unmarshal because the types do not match apparently so you would need to manually assign the value to the new struct.

**Comment by fgimenez** [comment]
> I think we might want to fix this only for the compat API and not regular inspect

Ok done, checked in our use case and works fine, also added a few unit tests. @Luap99 @mheon PTAL

**Comment by rhatdan** [comment]
@fgimenez Tests are failing.

**Comment by fgimenez** [comment]
> @fgimenez Tests are failing.

@rhatdan thanks for the heads up! ok that error should be fixed now, I see now this test failing https://github.com/containers/podman/pull/12174/checks?check_run_id=4149334267 not sure if it is related to the changes in the PR?

**Comment by rhatdan** [comment]
@Luap99 @jwhonce @mheon PTAL

File: pkg/api/handlers/compat/containers.go
Comment (Luap99) [inline pkg/api/handlers/compat/containers.go:647]: I think you also have to set the PrefixLen field in this struct, although I am not sure how we could get this value here if we do not change the type in libpod inspect as well.

Comment (fgimenez) [reply pkg/api/handlers/compat/containers.go:647]: Indeed, I assumed that the strings in the current SecondaryIPv6Addresses's `[]string` in libpod inspect represent actual addresses, not CIDRs, so the default value of PrefixLen of 0 should be ok, is this correct?

Comment (fgimenez) [reply pkg/api/handlers/compat/containers.go:647]: If the strings in libpod's inspect output can contain CIDRs then maybe we could try to parse them, split by `/` if present or something like that, wdyt?

Comment (fgimenez) [reply pkg/api/handlers/compat/containers.go]: > Indeed, I assumed that the strings in the current SecondaryIPv6Addresses's `[]string` in libpod inspect represent actual addresses, not CIDRs, so the default value of PrefixLen of 0 should be ok, is this correct?

Sorry, this was wrong, if SecondaryIPv6Addresses's `[]string` in libpod inspect represent actual addresses then PrefixLen should be 64, not 0 (32 for IPv4). I've pushed changes to fix it, PTAL.

**Comment by TomSweeneyRedHat** [comment]
Changes LGTM
one last failure on the "fedora-34 rootless host" test, I've restarted.

File: pkg/api/handlers/compat/containers.go
Comment (Luap99) [reply pkg/api/handlers/compat/containers.go]: I think this is not correct. You should have the actual ip/prefix for this container, hard coding it to 64 or 32 is wrong.
TBH looking at this I think your original patch works better. We are on 4.0 so breaking changes are possible. I do not think there many users of the secondary ip field anyway.
@mheon WDYT?

Comment (mheon) [reply pkg/api/handlers/compat/containers.go]: Changing the secondary IP field in `podman inspect` itself?

You're probably right that there are almost no users. I'm not strongly opposed to making a breaking change here.

**Comment by flouthoc** [comment]
@fgimenez Needs a small test for inspect. :) or you could test CI via adding `NO TESTS NEEDED` to your commit message and add test later when everything is green.

**Comment by fgimenez** [comment]
> @fgimenez Needs a small test for inspect. :) or you could test CI via adding `NO TESTS NEEDED` to your commit message and add test later when everything is green.

@flouthoc ok I've added a unit test for the changes, PTAL

**Comment by flouthoc** [comment]
@containers/podman-maintainers Could there be any issues with CI I think some unrelated tests are failing.

**Comment by mheon** [comment]
https://github.com/containers/podman/pull/12343 will fix CI once it merges, after that this will need to be rebased

**Comment by fgimenez** [comment]
> #12343 will fix CI once it merges, after that this will need to be rebased

cool thx, I see it is already merged, rebased this one

**Comment by Luap99** [comment]
@fgimenez Could you squash your commits into one please, otherwise LGTM.

**Comment by fgimenez** [comment]
> @fgimenez Could you squash your commits into one please, otherwise LGTM.

sure, done

**Review by Luap99 — APPROVED** [review]
LGTM

**Pull request merged**
